packages/frontend/component/src/ui/dnd/draggable.ts

```diff
@@ -0,0 +1,114 @@
+import { draggable } from '@atlaskit/pragmatic-drag-and-drop/element/adapter';
+import { useEffect, useMemo, useRef, useState } from 'react';
+import { flushSync } from 'react-dom';
+
+type DraggableGetFeedback = Parameters<
+  NonNullable<Parameters<typeof draggable>[0]['getInitialData']>
+>[0];
+
+type DraggableGet<T> = T | ((data: DraggableGetFeedback) => T);
+
+function draggableGet<T>(
+  get: T
+): T extends undefined
+  ? undefined
+  : T extends DraggableGet<infer I>
+    ? (args: DraggableGetFeedback) => I
+    : never {
+  if (get === undefined) {
+    return undefined as any;
+  }
+  return ((args: DraggableGetFeedback) =>
+    typeof get === 'function' ? (get as any)(args) : get) as any;
+}
+
+export interface DraggableOptions<
+  T extends Record<string, unknown> = Record<string, unknown>,
+> {
+  data?: DraggableGet<T>;
+  dataForExternal?: DraggableGet<{
+    [Key in
+      | 'text/uri-list'
+      | 'text/plain'
+      | 'text/html'
+      | 'Files'
+      // eslint-disable-next-line @typescript-eslint/ban-types
+      | (string & {})]?: string;
+  }>;
+  canDrag?: DraggableGet<boolean>;
+}
+
+export const useDraggable = <
+  T extends Record<string, unknown> = Record<string, unknown>,
+>(
+  getOptions: () => DraggableOptions<T> = () => ({}),
+  deps: any[] = []
+) => {
+  const [dragging, setDragging] = useState<boolean>(false);
+
+  // used to capture the dragging state when the drag is started
+  const [capturing, setCapturing] = useState<boolean>(false);
+  // this flag will be set to true when the `capturing` state is being used
+  const enableCapturingState = useRef<boolean>(false);
+
+  const dragRef = useRef<any>(null);
+  const dragHandleRef = useRef<any>(null);
+
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  const options = useMemo(getOptions, deps);
+
+  useEffect(() => {
+    if (!dragRef.current) {
+      return;
+    }
+
+    return draggable({
+      element: dragRef.current,
+      dragHandle: dragHandleRef.current ?? undefined,
+      canDrag: draggableGet(options.canDrag),
+      getInitialData: draggableGet(options.data),
+      getInitialDataForExternal: draggableGet(options.dataForExternal),
+      onDragStart: () => {
+        setDragging(true);
+        if (dragRef.current) {
+          dragRef.current.dataset['dragging'] = 'true';
+        }
+      },
+      onDrop: () => {
+        setDragging(false);
+        if (dragRef.current) {
+          delete dragRef.current.dataset['dragging'];
+        }
+      },
+      onGenerateDragPreview() {
+        if (dragRef.current) {
+          dragRef.current.dataset['capturing'] = 'true';
+        }
+
+        requestAnimationFrame(() => {
+          if (dragRef.current) {
+            delete dragRef.current.dataset['capturing'];
+          }
+        });
+        if (enableCapturingState) {
```

Comment:
Isn't this always `true`?

Reply:
```ts
const { dragRef, capturing } = useDraggable() // enableCapturingState = true
const { dragRef } = useDraggable() // enableCapturingState = false
```

Reply:
But `enableCapturingState` is a `MutableRefObject`, should it be `enableCapturingState.current`?

Reply:
my mistake